---
title: "Introduction to R Week 1 Notebook"
output:
  html_notebook: default
  pdf_document: default
---

# Introduction to Population Genetics in R
Based on the Population genetics workshop by John Novembre and the Introduction to R found at: https://speciationgenomics.github.io/R_introduction/. Much of the text and examples originated from these publicly-available sources and was modified to suit this class

*Modified by Priscilla Erickson, University of Richmond, August 2021*

### Welcome 

You are reading an R notebook. There is text that has instructions and information,and you can add and modify it. For example, you can answer questions by typing directly in the document: type your name below:

**(your name here)**

Then you will see "code chunks", which are demarcated by lines that look like this:

``` {r}
print("your code here")
```


Click on the green arrow at the far right of the code above. See what happens? R executed that code and returned a value. In this case, it just printed "your code here".

This exercise is going to expose you the utility of using R for basic statistical analyses by looking at human genetic data to see where differences lie in the human genome.  

### A brief introduction to R

Let's play around with some basic things that R can do. 

R can be a calculator: see what happens when you run the code below.

```{r}
2+5
3/100000
(29*2)-(17/8)
#note that R ignores things with a pound sign (#): the next line will not be calculated
#9178/38902

#You can use this function to add comments to your work, as you'll see throughout this document

```


You can also "save" particular values in R as *variables* by assigning them names with the `<-` symbol . Here, I'm making a variable called `y` that is equal to 17. I confirm the value of `y` by simply typing `y` (which returns 17) Then I can do things with the variable, like add or multiply it.

```{r}
y<-17
y
y*2
y+13
```

Now you try it: make a *variable* called `x` that equals the product of your birth month and birth year in the code chunk below, then confirm the value of it. Then divide `x` by 10

```{r}
#make a variable called x:

#confirm the value of x:

#divide x by 10:

```

You can also make variables that are text rather than number; we call these "character" variables and the characters that they hold need to be entered in quotation marks ("") (but note the name of the variable is *not* in quotation marks)

```{r}
professor<-"Dr. Erickson"
professor
```

Now you try it: make a variable called `name` that is your first name and confirm that it works

```{r}
#your code here

```

Most things you'll do in R are "functions" that you will execute on numbers or variables. 
Two simple mathematical functions are taking the square root of a variable (`sqrt()`) or taking the logarithm of a number or variable (`log10`). To use a function, you type its name, followed by the value to execute it on in parentheses. Note that you can execute  the function on a number  you insert, or a variable that you've previously saved (like "y")

```{r}
sqrt(7893)
log10(y)
```


You can also use R to do conditional (true/false) operations using the symbols =, > and <. To test whether two values are equal, you have to use two equal symbols ("=="). When you run this code in the console, R will return TRUE or FALSE for each condition - denoting whether the expressions are indeed true or false. Also note below that some lines of code have a hashtag (#) before them. This tells R that these are annotations; R does not run any lines that begin with #. Annotations are useful to keep track of what you are doing/thinking in your code.

```{r}
# are the values equal?
2 == 2
# is the first value greater than the second?
5 > 10
# is the first value less than or equal to the second?
9 <= 10
```

You can also do these operations on character variables. 
```{r}
#Is the professor's name Dr. Jones?
professor=="Dr. Jones"

professor
```

Now you try it: use the code block below to answer the questions with code

```{r}
#is x greater than 10000?


# is your name Darwin? (remember that you previously saved a variable with your name)


```

Calculations and conditionals are two of the most important R functions we will use and we'll see them come up many times in future weeks. There are many more useful things to learn, but we will introduce them as needed. Now lets move on to some biology.

We're now going to transition to using some of these tools to analyze human genetic data.

### Introductory terminology for our case study

- Single-nucleotide polymorphism (SNP):  A nucleotide basepair that is *polymorphic* (i.e. it has multiple types or *alleles* in the population)
- Allele:  A particular variant form of DNA  (e.g. A particular SNP may have the "A-T" allele in one DNA copy and "C-G" in another. In many cases we don't care about the precise base, so we might call these simply the *A1* and *A2* alleles, or the *A* or *a* alleles, or the *0* and *1* alleles.) 
- Genotype: The set of alleles carried by an individual (E.g. AA, AC, CC; or AA, Aa, and aa; or 0/0, 0/1, 1/1, depending on what we are calling each allele.

### The data-set and basic pre-processing

We will look at a global-scale sampling of human diversity. Basically researchers sequenced the genomes of thousands of people and identified their genotypes at many (millions) SNPs. We have filtered down the individuals to a set of 938 unrelated individuals. These individuals come from 52 different populations around the world. We have also extracted the  counts of the three possible genotypes at each genetic variant. The files with these genotype frequencies are your starting points. 

### Note about logistics

R can do a lot on its own, but people have written thousands of "libraries" that add additional functions for different purposes. We will use some of functions from the `data.table` and `ggplot2` and  libraries. `data.table` allows us to manipulate data organized in rows and columns, like an Excel spreadsheet.  `ggplot2` is the gold-standard for making beautiful graphs in R. Let's load them now.

```{r message=FALSE}
library(data.table)
library(ggplot2)
library(rmarkdown)
library(tinytex)
#setwd("~/Library/CloudStorage/Box-Box/BIO351_S23/Lab/Week 2- Human genetic data/")
```

### Initial view of the data
The data that we are using was downloaded from Github along with the workbook that you are reading right now.
We will use a function called `fread()` to read in the data table and we will save it as an object called `g`. The header=TRUE tells R to read the first row of data as the column names.

```{r}
g <- fread("H938_chr15.geno", header=TRUE)
```

It will be read in as a special object called a data.table in R.

You can see what a few lines of the data.table look like by by using the function `head()` and putting the name of the the table in between the parentheses. Try it below: 

```{r}

```

You should see that there are columns, each with distinct names.  

CHR  SNP A1 A2 nA1A1 nA1A2 nA2A2

- CHR: The chromosome number out of our 23 chromosomes.  In this case they are all from chromosome 15.
- SNP: The id of a SNP is a unique, universal identifier for a particular SNP. 
- A1: The rarer allele at the SNP (DNA base)
- A2: The more common allele (DNA base)
- nA1A1 : The number of A1/A1 homozygotes
- nA1A2 : The number of A1/A2 heterozygoes
- nA2A2 : The number of A2/A2 homozygotes

###  Looking at the size of the data.table

We can see how many rows and columns are in our data.table using the `dim()` function or the `nrow(`) function just for number of rows, which is usually more useful. Try both below

```{r}


```

**Question** How many rows are there, and what does each row represent?

**Answer here:** Type your answer here

### Subsetting the data

We can use the conditionals shown above to pull out rows of the data.table that meet certain conditions. to do that, we put the conditional statement inside square brackets [] immediately after the name of the data.table. For example, we can find all the SNPs that have an A as the major allele using this command (we have to put "A" in quotes because it is the value of a character variable rather than an numerical variable). We can save this subset as a new data.table called `d`. How many SNPs have an adenine as the major allele?

```{r}
#view some of the rows where A1 is A
g[A1=="A"]
#filter the data to only have rows where A1 is A and save it as a new object called d
d <- g[A1=="A"]
#see the dimensions to see how many rows were A
nrow(d)
```

Alternatively, you could combine these into a single command without saving "d" as an intermediate; here we are filtering g and then immediatetly calculating the number of rows:

```{r}
nrow(g[A1=="A"])
```

Now you try it. How would you answer these questions? Enter your code in the chunk below

```{r}
#how many rows have a "G" as allele 2 (A2)?

#How many rows have at least 100 A1A1 homozygotes?

```

### Calculate the number of counts at each locus

Next we will compute the total *number of observations* at each SNP by summing each of the three possible genotypes.  Here we will add a new column to our data table using a special notation. The blank space before the first comma tells R that we want to use all the rows of the data.table (we could use a conditional statement like above to only apply the calculation to some rows, though). The := says "assign a new column" to the data table, and the expression after the := tells R how to calculate the new column. Essentially we are creating a formula like you might do in Excel to add up several columns and save the values in new column colled `nObs` for number of observations.

```{r}
#calculate the total number of observations
g[ ,nObs := nA1A1 + nA1A2 + nA2A2]
#look at our updated table
g
```

Look at the new version of your dataframe `g` and confirm it has a new column called `nObs`.

**Question:** In your own words, what is nObs telling us?

**Answer here**


The `ggplot2` library has the ability to easily plot data with just a bit of code with the command `qplot`. If we give the `qplot()   ` function the name of a single column and tell it where that column came from, it will make a histogram of the data for that column. Let's try it:

```{r}
qplot(nObs, data = g)
```

**Question:** Describe the plot. What is the most common value of nObs? Why are some values lower, but none are higher? You may need to look back at the information above to remind yourself what is included in this dataset.

**Answer here:**

### Calculating genotype and allele frequencies

Let's move on to calculating genotype frequencies. The genotype frequency is how common a particular genotype is relative to the whole population (in other words, what proportion of the population has that genotoype out of all the possible genotypes?)  For the A1A1 homozygous genotype, we will refer to its frequency as p11 (shorthand for "proportion of A1A1 genotypes"). The frequencies of A1A2 and A2A2 will be p12 and p22 respectively. Complete the last two lines of code below and calculate new columns for the genotype frequencies for all 3 genotypes.

```{r}
#The equation for p11 is:
g[, p11 := nA1A1/nObs]
# calculate p12 here:

# calculate p22 here:

#look at your data and confirm that all three columns have been added and have values:

```

### Compute allele frequencies from genotype frequencies

We have genotype frequencies and now we need to calculate the *allele frequencies* of the A1 and A2 alleles. The allele frequency is how common each allele is in the population. This is like counting up all the alleles in a Hardy-Weinberg problem. The equations are given below.

```{r}
g[,p1 := p11 + 0.5*p12] 
g[,p2 := p22 + 0.5*p12]
```

**Question:**  Explain the equations for p1 and p2. Where did the 0.5 come from? 

**Answer here:**

And let's plot the frequency of the major allele (A2) vs the frequency of the minor allele (A1).  We can use the qplot() function again, but give it two variables and it will automatically make a scatter plot of them.

```{r}
qplot(p1, p2, data=g)
```

**Question:**  Describe this graph. What is the sum of any given x+y and why?

**Answer here:**

### Plotting genotype as a function of allele frequencies

Let's look at an initial plot of genotype vs allele frequencies.  First we need to transform our data into a "melted" version, which we will save as a new object called `g.melt`. This is tricky operation, so don't worry too much about what the code says right now. 

```{r}
g.melt<-melt(g, id.vars=c("SNP", "p1"), measure.vars=c("p11", "p12", "p22"), value.name="frequency", variable.name="genotype")
g.melt
```

**Question:** Look at the `g.melt` object. What did we just do here? The data are the same, but how has the organization changed? (hint: you might want to look at how many rows are in `g.melt` relative to `g`)

**Answer here:**

Now we are going to make a plot of genotype frequencies as a function of the allele frequencies. First, we have to tell ggplot what data to use (that's the `ggplot(g.melt)` part). Then we have to tell it to draw points with different x and y values (`geom_point`) and tell it what to use as our x and y variables. We are going to color-code by each genotype with the `color=genotype` argument. Notice how I've broken the code into multiple lines to make it easier to read here.

```{r}
ggplot(g.melt) + geom_point(aes(x = p1,
                               y = frequency, 
                               color = genotype))
```



**Question:** In your own words, what is this plot showing? What does each point represent? What are the x and y axes? What types of relationships do you see?

**Answer here:**


Under assumptions that there is no mutation, no natural selection, infinite population size, random mating and no migration, then the genotype frequencies will take on a simple relationship with the allele frequencies. We typically use *p *and *q* (where q=1-p) for the frequencies of allele 1 and 2, and present these *Hardy-Weinberg proportions* as: *p^2*, *2pq*, and *q^2*. 

**Question** If you look at data and see that it does not conform to Hardy-Weinberg expectations, what does that tell you?

**Answer here** 

Let's add to the plot lines that represent Hardy-Weinberg proportions on top of our data. Don't worry about the code here, just know that you can use `ggplot` to plot lines too!

```{r}
ggplot(g.melt)+
  geom_point(aes(x=p1,y=frequency,color=genotype))+ #this is the same plot as before
  stat_function(fun=function(p) p^2, geom="line", colour="red",size=2.5) + #adding the p^2 line
  stat_function(fun=function(p) 2*p*(1-p), geom="line", colour="green",size=2.5) +
  stat_function(fun=function(p) (1-p)^2, geom="line", colour="blue",size=2.5)
```

**Question** How do the data  visually compare to the Hardy-Weinberg expectations? Are any genotypes over-represented or under-represented?

**Answer here**


**Question** Propose a hypothesis for why the data might look the way they do

**Answer here**

***Please stop here and discuss your answers above with the instructor before moving on***


### Testing Hardy Weinberg

Now let's use Pearson's Chi-square-test (remember Pearson from our lecture? He was not the greatest guy but developed a number of statistical tests that we will be using this semester). Chi-squared is a basic statistical test that can be used to see if count data conform to a particular expectation. There's a good chance you did it to test for Hardy-Weinberg Equlibirum in a lab for Bio 202. The test is based on calculing the Chi-squared-test statistic, which is equal to *[(observed-expected)^2]/expected]*, summed across all categories of data. 

Lots of mathematical theory shows that this statistic follows an expected distribution, and so if it is very big or relative to the expected distribution, we can say that our data deviate from the expectation.

Here we compute a new column with the test statistic(`X2`) and obtain its associated p-value (using the `pchisq()` function) with 1 degree of freedom in a second column

```{r}
g[,X2 := (nA1A1-nObs*p1^2)^2 /(nObs*p1^2) + 
            (nA1A2-nObs*2*p1*p2)^2 / (nObs*2*p1*p2) + 
            (nA2A2-nObs*p2^2)^2 / (nObs*p2^2)]
g[,pval := 1-pchisq(X2,1)]
g
```

**Question:** Why did we have to add together three lines to calculate our chi-squared statistic (X2) in the first piece of code?

**Answer here:**

**Question:** What is a P-value? How do we usually interpret a P-value?

**Answer here:**

**Question** What is our null hypothesis for this test?

**Answer here**

### The problem of multiple testing

Let's look at the top few p-values:

```{r}
head(g$pval)
```

A p-value gives us the frequency at which we would observe our data if the null hypothesis is true.  If the data are relatively rare under the null (e.g. p-value < 5%), we reject the null hypothesis, and we would infer that the given SNP departs from Hardy-Weinberg expectations.  This is problematic here though. The problem is that we are testing many, many SNPs. Even if the null is universally true, 5% of our SNPs would be expected to be rejected.  This is called the multiple testing problem and is an idea we will revisit in the course.  As an example, if we have 50,000 SNPs, that all obey the null hypothesis, we would on average naively reject the null for ~2500 SNPs based on the p-values < 0.05.  Many of these SNPs would be false positives and not important to our analysis.  

First, let's see how many tests have p-values less than 0.05.   

```{r}
#write an expression here to find all the rows of g with p<0.05


#how many rows are there?


#what fraction of the total # of SNPs is this?


```

**Question**: Are there more SNPs than we'd expect on average given the total number of SNPs and a 5% rate of rejection under the null? 

**Answer here** 

### Plotting expected vs observed heterozygosity

To understand this more clearly, let's make a quick plot of the expected vs observed heterozygosity (Heterozygosity the proportion of heterozygotes). The observed heterozygosity is how many hets we see, the expected is the number that we expect based on HWE predictions. 

```{r}
g[,expected_hets:=2*p1*p2] #expected rate of heterozygotes under HWE
ggplot(g)+geom_point(aes(x=expected_hets, y=p12)) + geom_abline(intercept = 0,
                                                slope=1,
                                                color="red",
                                                size=1.5)
```

**Question:** What do you see? Are heterozygotes more common, less commmon, or as common as predicted? How do you interpret this result?

**Answer here**

### Discussion: Population subdivision and departures from Hardy-Weinberg expectations

Given the  data are from 52 sub-populations from around the globe, and alleles have some some probability of clustering within populations, a good working hypothesis for the deficiency of heterozygotes in this dataset is the presence of some population structure. *In other words, some genotypes may be more common in some sub-populations of humans, and other genotypes might be more common in another population*

**Question:** Explain a reason why you might see different genotypes at different frequencies within different human populations

**Answer here**


### Finding specific loci that are large departures from Hardy-Weinberg

Now, let's ask if we can find any loci that are wild departures from HW proportions.  These might be loci that cluster geographically in dramatic ways (such that they have few heterozygotes relative to expectations). To find these loci, we'll compute a number called *F*, which is the relative deficiency of heterozygotes [or *(expected hets-observed hets)/expected hets)*]


```{r}
g[,F := ((2*p1*p2)-p12) / (2*p1*p2)]
```

**Question:** Explain in your own words what the F equation above is doing. which part of the equation is referring to oberved heterozygotes and which is calculating expected?

**Answer here:**

Now let's make a plot of all the values of F, in order across the chromosome

```{r}
plot(g$F, xlab = "SNP number")
```

There are a few interesting SNPS that show either a very high or low *F* value. 

Let's extract the SNP id for the largest value by using the `order()` function to sort g by the value of F (will sort from lowest to highest) and using `tail()` to look at the last 5 rows of the data table.
```{r}
tail(g[order(F)])
```

**Question:** Which SNP has the highest value of F? In your own words, what does that high value of F mean? 

**Answer here**

**Question** Search for this SNP at https://www.ncbi.nlm.nih.gov/snp/. Click on the name of SNP (rsXXX), then scroll down and click the "Frequency" link. How does the frequency of this SNP differ between African and European samples?

**Answer here**

The SNP you identified is near a gene called SLC24A5. 

**Question:** Carry out a Google scholar search  for this gene using the term "positive selection" and see what you find. It's thought that the high F value observed here is because natural selection led to a "geographic clustering" of alleles in this gene region, meaning that people living in one region of the world tend to have one allele, and people living in another region of the world may tend to have different alleles.  Discuss with your partners why this might or might not make sense.  

**Answer here**

**Question** You've spent this lab identifying one of the most highly genetically differentiated regions of the human genome and learned a bit about its biological function. What does that tell you about overall genetic difference between human populations? 

**Answer here**


*When you reach this point in the document, please chat with Dr. Erickson about your conclusions, especially for the last two questions, before leaving lab for the day.*

### References

Li, Jun Z, Devin M Absher, Hua Tang, Audrey M Southwick, Amanda M Casto, Sohini Ramachandran, Howard M Cann, et al. 2008. “Worldwide Human Relationships Inferred from Genome-Wide Patterns of Variation.” Science 319 (5866): 1100–1104.

Pickrell, Joseph K, Graham Coop, John Novembre, Sridhar Kudaravalli, Jun Z Li, Devin Absher, Balaji S Srinivasan, et al. 2009. “Signals of Recent Positive Selection in a Worldwide Sample of Human Populations.” Genome Research 19 (5): 826–37.